File: 18-4sum/18-4sum.java
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
         Arrays.sort(nums);
    Set<List<Integer>> set = new HashSet<>();
    for(int i = 0; i < nums.length; i++){
        for(int j = i+1; j < nums.length; j++){
            int start = j+1;
            int end = nums.length-1;
            while(end > start){
                int sum = nums[i]+ nums[j] + nums[start] + nums[end];
                if(sum == target){
                    set.add(Arrays.asList(nums[i], nums[j], nums[start++], nums[end--]));
                }
                else if (sum > target) end --;
                else start++;
            }
        }
    }
      return new ArrayList<>(set);
        
    }
}

File: 18-4sum/18-4sum.cpp
class Solution {
public:
    
   
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        sort(nums.begin(),nums.end());
        vector<vector<int>>ans;
        set<vector<int>>s;
        int n =  nums.size();
        for(int i = 0;i<nums.size();i++)
        {
            long long sum1  = target- nums[i];
            for(int j = i+1;j<nums.size();j++)
            { 
                int start = j+1;
                int end = n-1;
                long long sum = sum1 - nums[j];
               // cout<<sum<<" "<<"Sum";
                while(start < end)
                {
                    long long new_sum =  nums[start] + nums[end];
                  //  cout<<new_sum<<"-"<<sum<<" ";
                    if(new_sum == sum )
                    {
                        s.insert({nums[i],nums[j],nums[start], nums[end]});
                        start++;
                        end--;
                    }
                    else if(new_sum > sum)
                    {
                        end--;
                    }
                    else
                    {
                        start++;
                    }
                    
                    
                    
                    
                }
                
            }
            
        }
        
        
        for(auto x:s)
        {
            ans.push_back(x);
        }
        return ans;
        
        
        
        
        
    }
};